acq_stat_reports/utils.py

```diff
@@ -0,0 +1,538 @@
+    sigma_1_high_frac[~sel] = np.nan
+    sigma_2_high_frac[~sel] = np.nan
+
+    (scatter,) = plt.plot(quantiles[col], acqid_frac, ".", color="k")
+
+    x = _mpl_hist_steps(
+        quantiles[f"{col}_low"],
+        quantiles[f"{col}_high"],
+    )
+    sigma_1_low = _mpl_hist_steps(sigma_1_low_frac)
+    sigma_2_low = _mpl_hist_steps(sigma_2_low_frac)
+    sigma_1_high = _mpl_hist_steps(sigma_1_high_frac)
+    sigma_2_high = _mpl_hist_steps(sigma_2_high_frac)
+
+    sigma_band_1 = plt.fill_between(
+        x, sigma_1_low, sigma_1_high, color="gray", alpha=0.8
+    )
+    sigma_band_2 = plt.fill_between(
+        x, sigma_2_low, sigma_2_high, color="gray", alpha=0.3
+    )
+
+    bins = bins[np.isfinite(bins)]
+    plt.xlim(bins[0], bins[-1])
+    plt.ylim(-0.05, 1.05)
+
+    labels = {
+        scatter: "Observed",
+        sigma_band_1: "68.2% range",
+        sigma_band_2: "95.4% range",
+    }
+    plt.legend(
+        labels.keys(),
+        labels.values(),
+        loc="best",
+    )
+
+
+@mpl_plot(
+    title="Acquisition Probability",
+    xlabel="Expected p$_{acq}$",
+    ylabel="Observed p$_{acq}$",
+    figscale=(1, 2),
+)
+def binned_data_probability_plot(
+    binned_data: BinnedData,
+    **kwargs,  # noqa: ARG001 (kwargs is neeeded by the decorator)
+):
+    quantiles = binned_data.binned_data
+
+    sel = quantiles["n"] > 0
+    prob_exp = quantiles["p_acq_model_mean"][sel]
+    prob_obs, low, high = binomial_confidence_interval(
+        quantiles["acqid"][sel], quantiles["n"][sel]
+    )
+
+    plt.gca().set_aspect("equal")
+    plt.errorbar(
+        prob_exp,
+        prob_obs,
+        yerr=[prob_obs - low, high - prob_obs],
+        xerr=quantiles["p_acq_model_std"][sel],
+        fmt=".",
+        color="k",
+        linewidth=1,
+    )
+    x = np.linspace(0, 1, 10)
+    plt.plot(x, x, "k--")
+    plt.ylim(0, 1)
+    plt.xlim(0, 1)
+
+
+def _get_quantiles_(p_acq_model, quantiles=None, n_realizations=10000):
+    """
+    Generate the 5th, 50th, and 95th percentiles of the expected number of successes in a series.
+
+    This function assumes the input is an array of probabilities, one for each draw.
+    It generates `n_realizations` realizations , and uses them to estimate the percentiles.
+
+    Parameters
+    ----------
+    p_acq_model : array-like
+        The binomial probability.
+    n_realizations : int, optional
+        The number of realizations of the series to use for the Monte Carlo estimation.
+
+    Returns
+    -------
+    array-like
+        The quantiles
+    """
+    if quantiles is None:
+        quantiles = [15.9, 50, 84.1]
+    samples = (
+        np.random.uniform(size=n_realizations * len(p_acq_model)).reshape(
+            (-1, len(p_acq_model))
+        )
+        < p_acq_model[None]
+    )
+    n = np.sum(samples, axis=1)
+    return np.percentile(n, quantiles)
+
+
+def get_histogram_quantile_ranges(  # noqa: PLR0915
+    data,
+    bin_edges,
+    extra_cols=(),
```

Comment:
Losely speaking, this function requires three basic ingredients, as you said in another comment: `successes`, `probs` and `bins`, but sometimes you want to get the mean, min and max for other quantities that are not directly related to the bins.

A typical example is plotting the mean `p_acq` for each magnitude bin. `p_acq` is not in the bin definition, so it does not appear in the result, but if you want to make that plot, you add it to `extra_cols`.

I have used it these days, it just didn't end up in a final plot.

Reply:
one convention could be to get the max, min and mean of all columns in the table, so the extra_cols argument is not needed. The the function calls need to be changed to only include the columns that we want in the result.

Reply:
> A typical example is plotting the mean p_acq for each magnitude bin. p_acq.

I think this would be best done with a call to `scipy.stats.binned_statistic`. The goal is to make functions that are minimal and do just one thing. 

But having now spent a huge amount of time staring at this function, I can at least see what `extra_cols` is doing. So if that is your preference then that's fine, but it needs to be documented carefully because the outputs are different.

Reply:
I made changed in [0a012a](https://github.com/sot/acq_stat_reports/pull/6/commits/0a012a7dfa0a2d34c23239efdc63bdfe7b1a1cb9) to make this function more readable. I'm not sure how `scipy.stats.binned_statistic` will make this significantly different from the latest version.